rawfile/filesystem/base.py

```diff
@@ -0,0 +1,82 @@
+from abc import ABC, ABCMeta, abstractmethod
+
+
+class NotSupportedError(NotImplementedError):
+    """
+    Exception raised when a feature is not supported by the current filesystem.
+    """
+
+    def __init__(self, feature: str) -> None:
+        super().__init__(
+            f"Feature '{feature}' is not supported by the current filesystem."
+        )
+
+
+class FileSystem(ABC, metaclass=ABCMeta):
+    def __init__(self, device: str, mountpoint: str | None = None):
+        """
+        Initialize the file system with the given device.
+        """
+        self.device = device
+        self.mountpoint = mountpoint
+
+    @classmethod
+    def __subclasshook__(cls, subclass):
+        """
+        Check if the subclass implements the required methods.
+        """
+        required_methods = [
+            "format_fs",
+            "mount",
+            "unmount",
+            "resize",
+        ]
+        for method in required_methods:
+            if not (hasattr(subclass, method) and callable(getattr(subclass, method))):
+                return False
+        return True
+
+    @abstractmethod
+    def create_snapshot(self):
+        """
+        Create a snapshot of the current state of the file system.
+        """
+        raise NotSupportedError("create_snapshot")
+
+    @abstractmethod
+    def delete_snapshot(self):
+        """
+        Delete the most recent snapshot of the file system.
+        """
+        raise NotSupportedError("delete_snapshot")
+
+    @abstractmethod
+    def format_fs(self, options: list[str] = []):
```

Comment:
What would format_fs do if we already have the filesystem mountpoint?
Ah because it can be None? So we should only allow format_fs is mountpoint is None

Reply:
Yeah, We should format before mount like what we have in mount utils in go

Reply:
format and mount has been added